compose/preview.c

```diff
@@ -0,0 +1,358 @@
+ * **Children**
+ *
+ * None.
+ *
+ * ## Data
+ * - #PreviewWindowData
+ *
+ * The Preview Window stores its data (#PreviewWindowData) in MuttWindow::wdata.
+ *
+ * ## Events
+ *
+ * Once constructed, it is controlled by the following events:
+ *
+ * | Event Type               | Handler                    |
+ * | :----------------------- | :------------------------- |
+ * | #NT_EMAIL (#NT_ENVELOPE) | preview_email_observer()   |
+ * | #NT_WINDOW               | preview_window_observer()  |
+ * | MuttWindow::recalc()     | preview_recalc()           |
+ * | MuttWindow::repaint()    | preview_repaint()          |
+ */
+
+#include "config.h"
+#include <stdbool.h>
+#include <stddef.h>
+#include "private.h"
+#include "mutt/lib.h"
+#include "mutt/logging2.h"
+#include "config/lib.h"
+#include "email/lib.h"
+#include "core/lib.h"
+#include "gui/lib.h"
+#include "gui/mutt_window.h"
+#include "attach/lib.h"
+#include "convert/lib.h"
+#include "expando/lib.h"
+#include "menu/lib.h"
+#include "attach_data.h"
+#include "shared_data.h"
+
+/**
+ * struct PreviewWindowData - Data to fill the Preview Window
+ */
+struct PreviewWindowData
+{
+  struct Email *email;    ///< Email being composed
+  int scroll_offset;      ///< Scroll offset
+  struct MuttWindow *win; ///< Window holding the message preview
+  bool more_content;      ///< Is there more content to scroll down to?
+};
+
+/**
+ * preview_wdata_free - Free the Preview Data - Implements MuttWindow::wdata_free() - @ingroup window_wdata_free
+ */
+void preview_wdata_free(struct MuttWindow *win, void **ptr)
+{
+  if (!ptr || !*ptr)
+    return;
+
+  FREE(ptr);
+}
+
+/**
+ * preview_wdata_new - Create new Preview Data
+ * @retval ptr New Preview Data
+ */
+struct PreviewWindowData *preview_wdata_new(void)
+{
+  struct PreviewWindowData *wdata = mutt_mem_calloc(1, sizeof(struct PreviewWindowData));
+
+  return wdata;
+}
+
+/**
+ * draw_preview - Write the message preview to the compose window
+ * @param win    Window to draw on
+ * @param wdata  Preview Window data
+ */
+static void draw_preview(struct MuttWindow *win, struct PreviewWindowData *wdata)
+{
+  struct Email *e = wdata->email;
+
+  FILE *fp = mutt_file_fopen(e->body->filename, "r");
+  if (!fp)
+  {
+    mutt_perror("%s", e->body->filename);
+    return;
+  }
+
+  mutt_window_clear(win);
+
+  // XXX(ds): I'm not particularly happy with this flag but so far I couldn't
+  //          think of a better way.
+  wdata->more_content = false;
+
+  // XXX(ds): We read the message from disk on every repaint. I _think_ this
+  //          is fine as a repaint should only happen when a new attachment is
+  //          added or the number of headers in the envelope change. Both not
+  //          something that happens on every email and even if, not many times.
+  //
+  //          The alternative would be to cache the content in PreviewWindowData
+  //          but that would be much more involved and could also get ugly with
+  //          large messages.
```

Comment:
it's generally going to be a small file that doesn't get reviewed.

**Emails** may get large when files are attached, but the message text is seldom large.
We can afford a little inefficiency here.